### ISSUE TITLE
strings cannot be assigned to Literals under certain circumstances

### ISSUE THREAD

**Issue by floxay** [microsoft/pyright#9205]
**Describe the bug**
Since `v1.1.384` sometimes strings are not assignable to Literals that would otherwise match them in sequence types(?).

For example if a `ConfigDict` is supplied to `model_config` in a Pydantic model above the potentially problematic attribute (`flags`) Pyright will error out, however if the `model_config` appears below `flags` there is no error using Pyright via CLI but VS Code (via Pylance) still reports the error no matter if it's before or after `flags`.

**Code or Screenshots**
This gives an error in both VSC (Pylance) and CLI:
```python
from collections.abc import Sequence
from typing import Literal

from pydantic import BaseModel, ConfigDict

ExistingFlags = Literal["foo", "bar"]


class Model(BaseModel):
    model_config = ConfigDict()

    flags: Sequence[ExistingFlags] = ["bar"]
```
The error:
```console
error: Type "list[str]" is not assignable to declared type "Sequence[ExistingFlags]"
    "list[str]" is not assignable to "Sequence[ExistingFlags]"
      Type parameter "_T_co@Sequence" is covariant, but "str" is not a subtype of "ExistingFlags"
        Type "str" is not assignable to type "ExistingFlags"
          "str" is not assignable to type "Literal['foo']"
          "str" is not assignable to type "Literal['bar']" (reportAssignmentType)
```

But this one only gives an error in VSC (Pylance):
```python

class Model(BaseModel):
    flags: Sequence[ExistingFlags] = ["bar"]

    model_config = ConfigDict()
```

Pydantic v2.9.2 (latest at the moment)
Pylance: v2024.10.100 (pre-release)
Pyright CLI: 1.1.384

**Comment by thomasgtaylor** [comment]
@floxay I believe the version number you meant was: https://github.com/microsoft/pyright/releases/tag/1.1.384

Similarly, we've started seeing occurrences of something that is likely caused by the same change:

```text
Pyright: Type "LiteralString" is not assignable to return type "str"
  "builtins.str" is not assignable to "builtins.str"
```

For example, `StrEnum` values are _now_ resolved as `LiteralString` when in f-strings:

```python
class SomeEnum(StrEnum):
    ONE = "one"

haiku: str = f"{ModelName.HAIKU.value}" # This does not work anymore

"""
Pyright: Type "LiteralString" is not assignable to declared type "str"
  "builtins.str" is not assignable to "builtins.str"
"""
```

**Comment by erictraut** [comment]
Thanks for the bug report.

This regression was specific to the case where a dataclass (or dataclass_transform class) has a default value expression that requires bidirectional type inference for the correct type evaluation. This specific combination of factors wasn't covered in pyright's unit tests, and it (somewhat surprisingly) didn't show up in the mypy_primer results.

This will be addressed in the next release, and I've added a test case to prevent a similar regression from going unnoticed in the future.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.385.
